prov/efa/src/efa_user_info.c

```diff
@@ -302,38 +303,58 @@ bool efa_user_info_should_support_hmem(int version, const struct fi_info *hints)
 	 */
 	if (FI_VERSION_GE(version, FI_VERSION(1, 18)) && hmem_ops[FI_HMEM_CUDA].initialized) {
 		EFA_INFO(FI_LOG_CORE,
-			"User is using API version >= 1.18. CUDA library and device are availble, claim support of FI_HMEM.");
+			"User is using API version >= 1.18. CUDA library and "
+			"device are availble, claim support of FI_HMEM.");
 		return true;
 	}
 
 	if (ofi_hmem_p2p_disabled()) {
 		EFA_WARN(FI_LOG_CORE,
 			"FI_HMEM capability currently requires peer to peer "
-			"support, which is disabled because FI_HMEM_P2P_DISABLED was set to 1/on/true.");
+			"support, which is disabled because "
+			"FI_HMEM_P2P_DISABLED was set to 1/on/true.");
 		return false;
 	}
 
-	if (!efa_device_support_rdma_read()) {
-		EFA_WARN(FI_LOG_CORE,
-			"FI_HMEM capability requires RDMA, which this device does not support.");
-		return false;
+	if (rxr_env_or_default_use_device_rdma()) {
+		/* support FI_HMEM because we can use RDMA. */
+		return true;
 	}
+	/* Otherwise, we cannot support FI_HMEM.  Log some relevant reason. */
 
-	if (!rxr_env_get_use_device_rdma()) {
+	if (!hmem_ops[FI_HMEM_CUDA].initialized) {
+		extra_info = "no CUDA library or device was found.";
+	}
+	else {
+		assert(!FI_VERSION_GE(version, FI_VERSION(1, 18)));
+		extra_info = "the application is using libfabric "
+			"API <1.18, so CUDA cannot be used.";
+	}
+
+	if (false == rxr_get_default_use_device_rdma()) {
```

Comment:
Do we really need `rxr_get_default_use_device_rdma` here? 

It seems, we just want to check whether device support rdma,  for that we can use `efa_device_support_rdma()`. Right?

Reply:
Frankly efa_device_support_rdma_read and rxr_get_default_use_device_rdma  are essentially redundant, but they use a different approach: rxr_get_default checks the vendor id and then double-checks the device caps, while efa_device_support_rdma_read just checks device caps only.  Personally I think the device-capabilities is the right way.

Should we remove get_default_use_device_rdma ?

Reply:
> Should we remove get_default_use_device_rdma ?

Yes. I think so.

Reply:
`get_default_use_device_rdma`  is gone.  `rxr_env_get_use_device_rdma` returns the environment value or the default, and a  new function (`rxr_env_is_set_use_device_rdma`) can be called to query if the environment variable is set.